src/report.rs

```diff
@@ -0,0 +1,237 @@
+    ///     Ok(42)
+    /// }
+    /// ```
+    pub fn capture_into_result<T>(body: impl FnOnce() -> Result<T, E>) -> Result<T, Self> {
+        body().map_err(Self::from_error)
+    }
+
+    /// Executes a closure that returns a [`Result`][], converting any
+    /// error to a [`Report`][].
+    ///
+    /// Recommended if you only support Rust version 1.61 or above.
+    ///
+    /// ```rust
+    /// use snafu::{prelude::*, Report};
+    ///
+    /// #[derive(Debug, Snafu)]
+    /// struct PlaceholderError;
+    ///
+    /// fn main() -> Report<PlaceholderError> {
+    ///     Report::capture(|| {
+    ///         let _v = may_fail_with_placeholder_error()?;
+    ///
+    ///         Ok(())
+    ///     })
+    /// }
+    ///
+    /// fn may_fail_with_placeholder_error() -> Result<u8, PlaceholderError> {
+    ///     Ok(42)
+    /// }
+    /// ```
+    pub fn capture(body: impl FnOnce() -> Result<(), E>) -> Self {
+        Self(body())
+    }
+
+    /// A [`Report`][] that indicates no error occurred.
+    pub const fn ok() -> Self {
+        Self(Ok(()))
+    }
+}
+
+impl<E> fmt::Debug for Report<E>
+where
+    E: crate::Error,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+impl<E> fmt::Display for Report<E>
+where
+    E: crate::Error,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.0 {
+            Err(e) => ReportFormatter(e).fmt(f),
+            _ => Ok(()),
+        }
+    }
+}
+
+#[cfg(all(feature = "std", feature = "rust_1_61"))]
+impl<E> Termination for Report<E>
+where
+    E: crate::Error,
+{
+    fn report(self) -> ExitCode {
+        match self.0 {
+            Ok(()) => ExitCode::SUCCESS,
+            Err(e) => {
+                eprintln!("{}", ReportFormatter(&e));
+
+                #[cfg(feature = "unstable-provider-api")]
+                {
+                    use core::any;
+
+                    any::request_value::<ExitCode>(&e)
+                        .or_else(|| any::request_ref::<ExitCode>(&e).copied())
+                        .unwrap_or(ExitCode::FAILURE)
+                }
+
+                #[cfg(not(feature = "unstable-provider-api"))]
+                {
+                    ExitCode::FAILURE
+                }
+            }
+        }
+    }
+}
+
+#[cfg(feature = "unstable-try-trait")]
+impl<T, E> core::ops::FromResidual<Result<T, E>> for Report<E> {
+    fn from_residual(residual: Result<T, E>) -> Self {
+        Self(residual.map(drop))
+    }
+}
+
+struct ReportFormatter<'a>(&'a dyn crate::Error);
+
+impl<'a> fmt::Display for ReportFormatter<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "{}", self.0)?;
+
+        let mut chain = ChainCompat::new(self.0).skip(1).peekable();
+        if chain.peek().is_some() {
+            writeln!(f, "")?;
+
+            for source in chain {
+                writeln!(f, "Caused by: {}", source)?;
+            }
```

Comment:
I'm down with that. I formatted the numbers as right-aligned with a width of 3, so it should look OK so long as you don't have 1000 nested errors...

Reply:
> ```
> Caused by:
>   0: could not connect to server
>   1: No route to host (os error 113)
> ```

I fear this leaves unclear the order in which the errors occurred.  Reading this output, I would think that error 0 led to error 1 (leading to error 2, and so on), but the "X, caused by Y, caused by Z" form makes clear that it's actually the reverse (error 1 led to error 0).

Reply:
> I fear this leaves unclear the order in which the errors occurred. Reading this output (for arbitrary errors), I would think that error 0 led to error 1 (leading to error 2, and so on), but the "X, caused by Y, caused by Z" form makes clear that it's actually the reverse (error 1 led to error 0).

Practically speaking, the order is the same. The most efficient and intuitive way of traversing the chain of sources is _outer to inner_, in which the root cause always appears last. All error reporting tools in Rust that I know of show the same order when printing an error report. These tools won't go out of their way to show this information in the opposite order.

Not to mention that the meaning of each error message is often enough to disambiguate the intended order of interpretation (e.g. "No route to host (os error #)" makes sense as a cause for not being able to connect to the server, whereas the opposite does not).

I disagree that this format is confusing.

Reply:
Recall that this cause trace won't be shown in isolation; it's right after the outermost error:

```
The highest level

Caused by:
  0: The middle level
  1: The lowest level
```

That's from this code:

```rust
#[derive(Debug, Snafu)]
#[snafu(display("The lowest level"))]
struct Inner;

#[derive(Debug, Snafu)]
#[snafu(display("The middle level"))]
struct Middle {
    source: Inner,
}

#[derive(Debug, Snafu)]
#[snafu(display("The highest level"))]
struct Outer {
    source: Middle,
}
```

If the trace _did_ go in the other direction, it'd result in an abrupt switch in direction:

```
The highest level

Caused by:
  0: The lowest level
  1: The middle level
```

---

These are all arguments for why the proposed formatting is "obviously correct", but I'm open to attempting to clarify the wording if possible. For example, since we only say it once, would expanding the "caused by" text be helpful at all? As a starting example:

```
The highest level

This was caused by these errors, in order:
  0: The middle level
  1: The lowest level
```

Reply:
>```
>This was caused by these errors, in order:
>  0: The middle level
>  1: The lowest level
>```

I still think "these errors, in order" suggests that the errors are listed in the order in which they were raised.  What about "This was caused by these earlier errors (with the earliest error listed last):"?

Reply:
> What about "This was caused by these earlier errors (with the earliest error listed last):"?

Seems reasonable; I picked a slight modification:`"Caused by these errors (earliest error listed last):"`.

I wonder if

1. we should negate the parenthetical... "latest error listed first".
2. we should do the right English thing and print "this error" when appropriate.

> can we also format it as

Yeah, I thought about that briefly when writing it, so I agree.